src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -1591,9 +1592,9 @@ public ValueTask<FlushResult> WritePipeAsync(ReadOnlyMemory<byte> data, Cancella
             VerifyAndUpdateWrite(data.Length);
         }
 
-        if (_canWriteResponseBody)
+        if (_responseBodyMode != ResponseBodyMode.Disabled)
         {
-            if (_autoChunk)
+            if (_responseBodyMode == ResponseBodyMode.Chunked)
```

Comment:
Nit: Can this be refactored into a 3-case switch statement? Or maybe 4-case switch statement for the hopefully unreachable `Uninitialized` if we decide add that.